src/algebra/big_operators/order.lean

```diff
@@ -308,3 +309,50 @@ open opposite
 (@unop_add_hom β _).map_sum _ _
 
 end with_top
+
+namespace finset
+
+/--
+The strong pigeonhole principle for finitely many pigeons and pigeonholes.
+-/
+lemma strong_pigeonhole {s : finset α} {t : finset β} (hne : t.nonempty) [decidable_eq β]
+  (f : α → β) (hf : ∀ a ∈ s, f a ∈ t)
+  (n : ℕ) (hn : t.card * n ≤ s.card) :
+  ∃ y ∈ t, n ≤ (s.filter (λ x, f x = y)).card :=
```

Comment:
I feel like this is not actually the strongest statement, and you should instead assume `t.card * n < s.card` and conclude `n < (s.filter (λ x, f x = y)).card`.

With this variant, you no longer need a `nonempty` assumption, either.

Reply:
Oh right... I intended for `n` to be the ceiling of `s.card / t.card` and kept assuming that's what I did!

(Convincing myself you have the strongest statement: for `n` to be the ceiling, it would be `hn : t.card * n ≤ s.card + t.card - 1`, and with `n = n' + 1`, this becomes `hn : t.card * n' ≤ s.card - 1`, which can safely be written as `hn : t.card * n' < s.card`.)

Reply:
Updated it to the stronger statement.  This makes it so `finset.univ_nonempty` is not used by this PR -- should I leave it in since it still seems useful to have around?